### ISSUE TITLE
Enhance no argument tool definition

### ISSUE THREAD

**Issue by Unshure** [strands-agents/sdk-typescript#280]
# Enhance no argument tool definition

## Overview

Currently, when trying to define a tool that takes no arguments, there is no intuitive way to make it work. Users must resort to workarounds like `z.object({ _: z.string().optional() })`, which is counter-intuitive and adds unnecessary complexity.

### Current Behavior

```typescript
// Attempted approach 1: Empty object (FAILS with JSON parse error)
const getPreferences = tool({
  name: 'get_preferences',
  description: 'Gets user preferences',
  inputSchema: z.object({}),  // ❌ SyntaxError: Unexpected end of JSON input
  callback: (input, context) => {
    return { preferences: getPrefs() }
  }
})

// Attempted approach 2: z.null() (FAILS with validation error)
inputSchema: z.null()  // ❌ Error: must be one of type: object

// Current workaround (works but awkward)
inputSchema: z.object({ _: z.string().optional() })  // ✓ Works
```

### JSON Parse Error
The "Unexpected end of JSON input" error occurs in `src/models/model.ts` at line 200:
```typescript
input: JSON.parse(accumulatedToolInput),
```

**Error flow:**
1. Tool defined with `z.object({})` generates valid schema: `{ type: 'object', properties: {}, additionalProperties: false }`
2. Model interprets this and sends tool use with empty string input: `""`
3. `JSON.parse("")` throws `SyntaxError: Unexpected end of JSON input`
4. Error is caught and logged but tool execution fails

### Why z.null() Doesn't Work
- Bedrock API requires `inputSchema.json.type` to be `"object"`
- `z.null()` generates `{ type: "null" }` which is rejected by Bedrock validation

### Why z.undefined() and z.void() Don't Work
- Zod throws errors when trying to convert these to JSON Schema:
  - `z.undefined()`: "Undefined cannot be represented in JSON Schema"
  - `z.void()`: "Void cannot be represented in JSON Schema"

### 1. Make inputSchema Optional

Make `inputSchema` optional in the following interfaces:

**`src/tools/types.ts`** - ToolSpec interface:
```typescript
export interface ToolSpec {
  name: string
  description: string
  inputSchema?: JSONSchema  // Make optional
}
```

**`src/tools/function-tool.ts`** - FunctionToolConfig interface:
```typescript
export interface FunctionToolConfig {
  name: string
  description: string
  inputSchema?: JSONSchema  // Make optional
  callback: FunctionToolCallback
}
```

**`src/tools/zod-tool.ts`** - ToolConfig interface:
```typescript
export interface ToolConfig<TInput extends z.ZodType, TReturn extends JSONValue = JSONValue> {
  name: string
  description?: string
  inputSchema?: TInput  // Make optional
  callback: (input: z.infer<TInput>, context?: ToolContext) => ...
}
```

### 2. Support z.void() as inputSchema

When `inputSchema` is `z.void()`:
- Detect using `instanceof ZodVoid` (before calling `z.toJSONSchema()`)
- Treat the same as when `inputSchema` is undefined
- Generate default empty object schema internally

### 3. Default Schema Generation

When `inputSchema` is undefined or `z.void()`:
- Generate default schema: `{ type: 'object', properties: {}, additionalProperties: false }`
- This satisfies Bedrock's requirement for `type: 'object'`
- Allows model to understand no input is expected

### 4. Fix JSON.parse Bug

In `src/models/model.ts` around line 200, handle empty input gracefully:
```typescript
// Before (fails with empty string)
input: JSON.parse(accumulatedToolInput),

// After (handles empty string)
input: JSON.parse(accumulatedToolInput || '{}'),
```

### 5. Callback Signature Design Decision

**User Question:** Can callback signature be `() => {...}` without input parameter?

**Decision:** Keep consistent callback signature with input parameter always present:
- Signature remains: `callback: (input, context?) => result`
- When `inputSchema` is undefined/void, input is typed as `Record<string, never>`
- Users can ignore the parameter: `callback: (_input, context) => result`

**Rationale:**
- Simpler implementation (avoids complex conditional TypeScript types)
- Maintains consistency across all tools
- Slightly more verbose but clearer intent

### For Zod Tools (src/tools/zod-tool.ts)

In the `ZodTool` constructor:
```typescript
constructor(config: ToolConfig<TInput, TReturn>) {
  super()
  const { name, description = '', inputSchema, callback } = config

  this._inputSchema = inputSchema
  this._callback = callback

  let generatedSchema: JSONSchema
  
  // Handle undefined or z.void()
  if (!inputSchema || inputSchema instanceof ZodVoid) {
    generatedSchema = {
      type: 'object',
      properties: {},
      additionalProperties: false
    }
  } else {
    // Generate JSON Schema from Zod and strip $schema property
    const schema = z.toJSONSchema(inputSchema) as JSONSchema & { $schema?: string }
    const { $schema, ...schemaWithoutMeta } = schema
    generatedSchema = schemaWithoutMeta as JSONSchema
  }

  // Create FunctionTool with generated schema
  this._functionTool = new FunctionTool({
    name,
    description,
    inputSchema: generatedSchema,
    callback: (input: unknown, toolContext: ToolContext) => {
      // Only validate if schema was provided
      const validatedInput = inputSchema ? inputSchema.parse(input) : input
      return callback(validatedInput, toolContext) as ...
    },
  })
}
```

### For FunctionTool (src/tools/function-tool.ts)

In the `FunctionTool` constructor:
```typescript
constructor(config: FunctionToolConfig) {
  super()
  this.name = config.name
  this.description = config.description
  
  // Use provided schema or default empty object schema
  const inputSchema = config.inputSchema ?? {
    type: 'object',
    properties: {},
    additionalProperties: false
  }
  
  this.toolSpec = {
    name: config.name,
    description: config.description,
    inputSchema: inputSchema,
  }
  this._callback = config.callback
}
```

## Files to Modify

1. **src/tools/types.ts**
   - Make `ToolSpec.inputSchema` optional

2. **src/tools/function-tool.ts**
   - Make `FunctionToolConfig.inputSchema` optional
   - Add default schema when undefined
   - Update TSDoc documentation

3. **src/tools/zod-tool.ts**
   - Make `ToolConfig.inputSchema` optional
   - Handle `z.void()` detection
   - Add default schema when undefined or void
   - Update TSDoc documentation

4. **src/models/model.ts**
   - Fix JSON.parse to handle empty string: `JSON.parse(accumulatedToolInput || '{}')`

5. **Tests**
   - `src/tools/__tests__/tool.test.ts` - Test undefined inputSchema
   - `src/tools/__tests__/zod-tool.test.ts` - Test undefined and z.void() inputSchema
   - Update existing tests if needed

## Acceptance Criteria

- [ ] `inputSchema` can be omitted in `tool()` function
- [ ] `inputSchema` can be `z.void()` in `tool()` function  
- [ ] `inputSchema` can be omitted in `FunctionTool` constructor
- [ ] When omitted/void, tools work correctly with models (Bedrock and OpenAI)
- [ ] JSON.parse error is fixed for empty input strings
- [ ] Callback receives empty object `{}` when no schema provided
- [ ] TypeScript types correctly infer input as `Record<string, never>` when no schema
- [ ] All existing tests pass
- [ ] New tests cover undefined and z.void() scenarios
- [ ] TSDoc documentation updated for affected interfaces
- [ ] No breaking changes to existing tool definitions

## Example Usage After Implementation

```typescript
// Approach 1: Omit inputSchema entirely (recommended)
const getPreferences = tool({
  name: 'get_preferences',
  description: 'Gets user preferences',
  // inputSchema omitted - no input required
  callback: (_input, context) => {
    const userId = context?.agent.state.get('userId') as string
    return { preferences: database.getPreferences(userId) }
  }
})

// Approach 2: Use z.void() (explicit intent)
const getStatus = tool({
  name: 'get_status',
  description: 'Gets current system status',
  inputSchema: z.void(),  // Explicitly no input
  callback: (_input, context) => {
    return { status: 'operational', uptime: getUptime() }
  }
})

// Approach 3: FunctionTool with optional schema
const refreshCache = new FunctionTool({
  name: 'refresh_cache',
  description: 'Refreshes the cache',
  // inputSchema omitted
  callback: (_input, toolContext) => {
    cache.refresh()
    return 'Cache refreshed'
  }
})
```

## Estimated Scope

- **Complexity**: Low to Medium
- **Files Modified**: ~4 source files + tests
- **New Tests**: 4-6 test cases
- **Breaking Changes**: None
- **Time Estimate**: 1-2 days

### Why Not Support Variable Callback Signatures?

The user asked about supporting `callback: (context) => result` when no input is needed. While this would be more ergonomic, it would require:
- Complex conditional TypeScript types
- Different runtime behavior based on schema presence
- More complex implementation and maintenance

The decision to keep a consistent callback signature prioritizes simplicity and maintainability over minor ergonomic improvements.

### Repository Documentation Updates

Consider updating these files to clarify this feature:
- `README.md` - Add example of tool without input
- `AGENTS.md` - Document no-input tool pattern
